src/libraries/System.Net.Http/src/System/Net/Http/HttpRequestMessage.cs

```diff
@@ -183,7 +185,28 @@ private void InitializeValues(HttpMethod method, Uri? requestUri)
 
         internal bool MarkAsSent()
         {
-            return Interlocked.Exchange(ref _sendStatus, MessageAlreadySent) == MessageNotYetSent;
+            return Interlocked.CompareExchange(ref _sendStatus, MessageAlreadySent, MessageNotYetSent) == MessageNotYetSent;
+        }
+
+        internal void MarkAsTrackedByTelemetry()
+        {
+            Debug.Assert(_sendStatus != MessageShouldEmitTelemetry);
+            _sendStatus = MessageShouldEmitTelemetry;
+        }
+
+        internal void OnAborted() => OnStopped(aborted: true);
+
+        internal void OnStopped(bool aborted = false)
+        {
+            if (_sendStatus == MessageShouldEmitTelemetry && Interlocked.Exchange(ref _sendStatus, MessageAlreadySent) == MessageShouldEmitTelemetry)
```

Comment:
I assume that is to help out the linker?

Otherwise, the conditon can only hold if we've already checked `IsEnabled()`.

Reply:
correct, the illinker won't be able to detect that

Reply:
Won't it lead to reporting higher number of active requests in case the telemetry is stopped in the middle of the request? Which is something that's being fixed by this PR.

Reply:
The PR is addressing the race condition in the case where Telemetry is enabled when requests have already been started.

The scenario where Telemetry is switched off and then re-enabled is considered a [known race-condition](https://github.com/dotnet/runtime/pull/37619#discussion_r444580315).